File: frontend/src/mock/channels/screenshot.ts
/**
 * `screenshot.*` — read where the game wrote them, so the mock keeps a few
 * rows and deletes from that list.
 */
import type { Screenshot } from '@/api/types';

import * as entries from '../state/entries';
import { type Handlers, str } from '../support';

let shots: Screenshot[] = entries
  .listInstances()
  .flatMap((instance, index: number) =>
    ['2026-09-01_18.42.10.png', '2026-08-30_21.05.55.png'].map(
      (file, shot) => ({
        instance: instance.id,
        instanceName: instance.name,
        file,
        path: `/home/you/.hestia/instances/${instance.id}/data/screenshots/${file}`,
        takenUnix: 1_757_000_000 - index * 86_400 - shot * 3_600,
        bytes: 1_482_000 + shot * 90_000,
      }),
    ),
  );

export const channels: Handlers = {
  'screenshot.list': (payload) => {
    const instance = str(payload, 'instance');
    return {
      screenshots: instance
        ? shots.filter(
            (shot) =>
              shot.instance === instance || shot.instanceName === instance,
          )
        : shots,
    };
  },
  'screenshot.delete': (payload) => {
    const instance = str(payload, 'instance');
    const file = str(payload, 'file');
    shots = shots.filter(
      (shot) => !(shot.instance === instance && shot.file === file),
    );
    return { screenshots: shots };
  },
};

File: crates/cli/src/commands/screenshot.rs
//! `hestia screenshot …` — the shots every instance took, read where they are.

use anyhow::Result;
use clap::Subcommand;

use crate::commands::mc::age_label;
use crate::ui::{self, View};

#[derive(Subcommand)]
pub enum ScreenshotCmd {
    /// Every screenshot the instances share
    #[command(alias = "ls")]
    List {
        /// One instance by name or id, whether or not it shares them
        #[arg(long)]
        instance: Option<String>,
    },
    /// Delete one screenshot from the instance that took it
    #[command(alias = "rm")]
    Remove {
        /// Instance name or id
        instance: String,
        /// The file, as `screenshot list` shows it
        file: String,
    },
}

pub async fn run(cmd: ScreenshotCmd) -> Result<()> {
    let client = super::connect().await?;
    match cmd {
        ScreenshotCmd::List { instance } => {
            let taken = client
                .screenshot()
                .list(instance.as_deref().unwrap_or_default())
                .await?;
            if taken.is_empty() {
                return ui::show(View::note(
                    "no screenshots yet — `hestia sync on screenshots` reads every instance's",
                ));
            }
            let rows = taken
                .into_iter()
                .map(|shot| {
                    vec![
                        age_label(shot.taken_unix),
                        shot.instance_name,
                        shot.file,
                        ui::human_bytes(shot.bytes),
                    ]
                })
                .collect();
            ui::show(View::table(
                "Screenshots",
                ["TAKEN", "INSTANCE", "FILE", "SIZE"],
                rows,
            ))
        }
        ScreenshotCmd::Remove { instance, file } => {
            client.screenshot().delete(&instance, &file).await?;
            ui::show(View::line(format!("deleted '{file}' from '{instance}'")))
        }
    }
}
